src/hotspot/os/linux/os_linux.cpp

```diff
@@ -194,18 +194,21 @@ julong os::Linux::available_memory() {
   julong avail_mem;
 
   if (OSContainer::is_containerized()) {
-    jlong mem_limit, mem_usage;
-    if ((mem_limit = OSContainer::memory_limit_in_bytes()) < 1) {
-      log_debug(os, container)("container memory limit %s: " JLONG_FORMAT ", using host value",
-                             mem_limit == OSCONTAINER_ERROR ? "failed" : "unlimited", mem_limit);
+    jlong mem_limit, mem_usage, host_mem;
+    host_mem = Linux::physical_memory();
+    mem_limit = OSContainer::memory_limit_in_bytes();
```

Comment:
With your latest changes in `cgroupV1Subsystem_linux.cpp` `mem_limit` would only exceed `host_mem` on cg2 with a memory limit specified `> host_mem` (very unlikely). Now we only do the gymnastics here to correct this situation. Instead, we could do it for both v1 and v2 in [CgroupSubsystem::memory_limit_in_bytes()](https://github.com/openjdk/jdk/blob/cf0067741249cc3260b1d220769dac408b614f22/src/hotspot/os/linux/cgroupSubsystem_linux.cpp#L525).

I'm not a big fan of this hybrid approach: correct cg1 values in `cgroupV1Subsystem_linux.cpp` and cg2 values in `os_linux`. My strong preference would be to do this in `CgroupSubsystem` code and keep `os_linux` code simpler.

Reply:
Thanks for explaining your opinion so clearly (here and on IRC). I've taken your advice and moved the sanity checking to `CgroupSubsystem` in bc493c20ba2cd0cd4044e4aa2839abd8589c403a. I think this looks and functions much better now.